### ISSUE TITLE
Add optional package cvxpy, update cylp, add CVXPY MIP backend

### ISSUE THREAD

**Issue by mkoeppe** [sagemath/sage#34251]
<div id="comment:0"></div>

(split out from #31962)

We add `cvxpy` for disciplined convex programming and extensions as an optional package.

- cvxpy ([home](https://www.cvxpy.org/) · [repo](https://github.com/cvxpy/cvxpy) · [releases](https://github.com/cvxpy/cvxpy/releases) · [PyPI](https://pypi.org/project/cvxpy/))
- It can call [many backend solvers](https://www.cvxpy.org/tutorial/advanced/index.html#choosing-a-solver) -- including most that we currently have as MIP backends, so it may be suitable as a replacement for our homegrown backend code
- Promising route to [add support for the HiGHS LP/MIP solvers](https://github.com/cvxpy/cvxpy/issues/1443) via the [scipy cython interface to their vendored HiGHS](https://github.com/scipy/scipy/issues/14455)
- supports [differentiable programming](https://www.cvxpy.org/tutorial/advanced/index.html#sensitivity-analysis-and-gradients)
- PR: https://github.com/cvxpy/cvxpy/pull/1703 - Work around broken CyLP infeasibility detection

We add a new MIP backend `solver="CVXPY"`. Using `solver="CVXPY/ECOS"`, `solver="CVXPY/GLPK"`, etc., CVXPY can be asked to use a specific solver.

When also `cylp` is installed (from #33487, upgraded here), the new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available), thus replacing the similar functionality of the CBC backend provided by `sage_numerical_backend_coin`.

Details about dependencies: Requirements listed in
cvxpy's [setup.py](https://github.com/cvxpy/cvxpy/blob/master/setup.py):

```
python_requires='>=3.6',
install_requires=["osqp >= 0.4.1",
                  "ecos >= 2",
                  "scs >= 1.1.6",
                  "numpy >= 1.15",
                  "scipy >= 1.1.0"],
```

So the new dependencies are:

- osqp (operator splitting quadratic program)  ([home](https://osqp.org/) · [repo](https://github.com/osqp/osqp-python) · [PyPI](https://pypi.org/project/osqp/))
- ecos (embedded conic solver)  ([repo](https://github.com/embotech/ecos-python) · [PyPI](https://pypi.org/project/ecos/))
- SCS (splitting conic solver)  ([repo](https://github.com/bodono/scs-python) · [PyPI](https://pypi.org/project/scs/))

In addition, osqp's [requirements.txt](https://github.com/osqp/osqp-python/blob/master/requirements.txt) lists:
`numpy >= 1.7`, `scipy >= 0.13.2`, `qdldl`

which adds:

- qldl-python (interface to [QDLDL](https://github.com/osqp/qdldl), to factor quasi-definite linear systems) ([repo](https://github.com/osqp/qdldl-python) · [PyPI](https://pypi.org/project/qdldl/))

Finally, qldl-python depends on pybind11 (already an SPKG)

See also:
- #26511 Meta-ticket: Use Python optimization interfaces: PuLP, Pyomo, cylp...
- #33493 package ortools - provides some more backends

Follow-ups:
- #30644 Remove `sage_numerical_backends_coin`, Upgrade optional package cbc to 2.10.7

Depends on #34228

CC:  @dimpase @yuan-zhou @sheerluck @dcoudert @seblabbe @maxale @kiwifb @isuruf @tkralphs

Component: **packages: optional**

Author: **Matthias Koeppe, Andrey Belgorodski**

Reviewer: **Martin Pépin**

_Issue created by migration from https://trac.sagemath.org/ticket/34251_

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1 +1,2 @@
 (split out from #31962)
+
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,2 +1,60 @@
 (split out from #31962)
 
+We add `cvxpy` for disciplined convex programming and extensions as an optional package.
+
+- cvxpy ([home](https://www.cvxpy.org/)
+  · [repo](https://github.com/cvxpy/cvxpy)
+  · [releases](https://github.com/cvxpy/cvxpy/releases)
+  · [PyPI](https://pypi.org/project/cvxpy/))
+- It can call [many backend solvers](https://www.cvxpy.org/tutorial/advanced/index.html#choosing-a-solver) -- including most that we currently have as MIP backends, so it may be suitable as a replacement for our homegrown backend code
+- Promising route to [add support for the HiGHS LP/MIP solvers](https://github.com/cvxpy/cvxpy/issues/1443) via the [scipy cython interface to their vendored HiGHS](https://github.com/scipy/scipy/issues/14455)
+- supports [differentiable programming](https://www.cvxpy.org/tutorial/advanced/index.html#sensitivity-analysis-and-gradients)
+- PR: https://github.com/cvxpy/cvxpy/pull/1703 - Work around broken CyLP infeasibility detection
+
+When `cylp` is installed (from #33487, upgraded here), a new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available). This replaces 
+
+
+Details about dependencies: Requirements listed in
+cvxpy's [setup.py](https://github.com/cvxpy/cvxpy/blob/master/setup.py):
+
+```
+python_requires='>=3.6',
+install_requires=["osqp >= 0.4.1",
+                  "ecos >= 2",
+                  "scs >= 1.1.6",
+                  "numpy >= 1.15",
+                  "scipy >= 1.1.0"],
+```
+
+So the new dependencies are:
+
+- osqp (operator splitting quadratic program)
+  ([home](https://osqp.org/)
+  · [repo](https://github.com/osqp/osqp-python)
+  · [PyPI](https://pypi.org/project/osqp/))
+- ecos (embedded conic solver)
+  ([repo](https://github.com/embotech/ecos-python)
+  · [PyPI](https://pypi.org/project/ecos/))
+- SCS (splitting conic solver)
+  ([repo](https://github.com/bodono/scs-python)
+  · [PyPI](https://pypi.org/project/scs/))
+
+In addition, osqp's [requirements.txt](https://github.com/osqp/osqp-python/blob/master/requirements.txt) lists:
+`numpy >= 1.7`, `scipy >= 0.13.2`, `qdldl`
+
+which adds:
+
+- qldl-python (interface
+  to [QDLDL](https://github.com/osqp/qdldl), to factor quasi-definite linear systems)
+  ([repo](https://github.com/osqp/qdldl-python)
+  · [PyPI](https://pypi.org/project/qdldl/))
+
+Finally, qldl-python depends on pybind11 (already an SPKG)
+
+See also:
+- #26511 Meta-ticket: Use Python optimization interfaces: PuLP, Pyomo, cylp...
+- #33493 package ortools - provides some more backends
+
+Follow-ups:
+- #30644 Remove `sage_numerical_backends_coin`, Upgrade optional package cbc to 2.10.7
+
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -11,7 +11,7 @@
 - supports [differentiable programming](https://www.cvxpy.org/tutorial/advanced/index.html#sensitivity-analysis-and-gradients)
 - PR: https://github.com/cvxpy/cvxpy/pull/1703 - Work around broken CyLP infeasibility detection
 
-When `cylp` is installed (from #33487, upgraded here), a new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available). This replaces 
+When `cylp` is installed (from #33487, upgraded here), a new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available). This replaces the backend provided by `sage_numerical_backend_coin`
 
 
 Details about dependencies: Requirements listed in
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -11,7 +11,9 @@
 - supports [differentiable programming](https://www.cvxpy.org/tutorial/advanced/index.html#sensitivity-analysis-and-gradients)
 - PR: https://github.com/cvxpy/cvxpy/pull/1703 - Work around broken CyLP infeasibility detection
 
-When `cylp` is installed (from #33487, upgraded here), a new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available). This replaces the backend provided by `sage_numerical_backend_coin`
+We add a new MIP backend `solver="CVXPY"`. Using `solver="CVXPY/ECOS"`, `solver="CVXPY/GLPK"`, etc., CVXPY can be asked to use a specific solver.
+
+When also `cylp` is installed (from #33487, upgraded here), the new solver backend `"CVXPY/CBC"` becomes the default (unless Gurobi or CPLEX are available), thus replacing the similar functionality of the CBC backend provided by `sage_numerical_backend_coin`.
 
 
 Details about dependencies: Requirements listed in
``````

**Comment by Kerl13** [comment]
<div id="comment:36" align="right">comment:36</div>

Some test do not pass, most notably: 

```
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/numerical/backends/cvxpy_backend.pyx
    Killed due to segmentation fault
**********************************************************************
Tests run before process (pid=287829) failed:
sage: import cvxpy ## line 42 ##
sage: cvxpy.installed_solvers()                                                # random ## line 43 ##
['CVXOPT', 'ECOS', 'ECOS_BB', 'GLPK', 'GLPK_MI', 'OSQP', 'SCIPY', 'SCS']
sage: p = MixedIntegerLinearProgram(solver="CVXPY"); p.solve() ## line 47 ##
0.0
sage: p = MixedIntegerLinearProgram(solver="CVXPY/OSQP"); p.solve() ## line 52 ##
0.0
sage: p = MixedIntegerLinearProgram(solver="CVXPY/ECOS"); p.solve() ## line 54 ##
0.0
sage: p = MixedIntegerLinearProgram(solver="CVXPY/SCS"); p.solve() ## line 56 ##
0.0
sage: p = MixedIntegerLinearProgram(solver="CVXPY/SciPy/HiGHS"); p.solve() ## line 58 ##
0.0
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 90 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 98 ##
sage: p = get_solver(solver="CVXPY") ## line 99 ##
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 100 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 141 ##
sage: p = MixedIntegerLinearProgram(solver="CVXPY") ## line 142 ##
sage: b = p.new_variable() ## line 143 ##
sage: p.add_constraint(b[1] + b[2] <= 6) ## line 144 ##
sage: p.set_objective(b[1] + b[2]) ## line 145 ##
sage: cp = copy(p.get_backend()) ## line 146 ##
sage: cp.solve() ## line 147 ##
0
sage: cp.get_objective_value()  # abs tol 1e-7 ## line 149 ##
5.999999982049109
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 151 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 206 ##
sage: p = get_solver(solver="CVXPY") ## line 207 ##
sage: p.ncols() ## line 208 ##
0
sage: p.add_variable() ## line 210 ##
0
sage: p.ncols() ## line 212 ##
1
sage: p.add_variable(continuous=True, integer=True) ## line 214 ##
sage: p.add_variable(name='x',obj=1) ## line 218 ##
1
sage: p.col_name(1) ## line 220 ##
'x'
sage: p.objective_coefficient(1) ## line 222 ##
1.0
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 224 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 287 ##
sage: p = get_solver(solver="CVXPY") ## line 288 ##
sage: p.set_verbosity(2) ## line 289 ##
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 290 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 313 ##
sage: p = get_solver(solver="CVXPY") ## line 314 ##
sage: p.add_variables(5) ## line 315 ##
4
sage: index = p.nrows() ## line 317 ##
sage: p.add_linear_constraint( zip(range(5), range(5)), 2, 2) ## line 318 ##
sage: p.row(index) ## line 319 ##
([1, 2, 3, 4], [1, 2, 3, 4])
sage: p.row_bounds(index) ## line 321 ##
(2, 2)
sage: p.add_linear_constraint( zip(range(5), range(5)), 1, 1, name='foo') ## line 323 ##
sage: p.row_name(1) ## line 324 ##
'constraint_1'
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 326 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 375 ##
sage: p = get_solver(solver="CVXPY") ## line 376 ##
sage: p.ncols() ## line 377 ##
0
sage: p.nrows() ## line 379 ##
0
sage: p.add_linear_constraints(5, 0, None) ## line 381 ##
sage: p.add_col(list(range(5)), list(range(5))) ## line 382 ##
sage: p.nrows() ## line 386 ##
5
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 388 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 412 ##
sage: p = get_solver(solver="CVXPY") ## line 413 ##
sage: p.add_variables(5) ## line 414 ##
4
sage: p.set_objective([1, 1, 2, 1, 3]) ## line 416 ##
sage: [p.objective_coefficient(x) for x in range(5)] ## line 417 ##
[1.0, 1.0, 2.0, 1.0, 3.0]
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 419 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 444 ##
sage: p = get_solver(solver="CVXPY") ## line 445 ##
sage: p.is_maximization() ## line 446 ##
True
sage: p.set_sense(-1) ## line 448 ##
sage: p.is_maximization() ## line 449 ##
False
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 451 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 472 ##
sage: p = get_solver(solver="CVXPY") ## line 473 ##
sage: p.add_variable() ## line 474 ##
0
sage: p.objective_coefficient(0) ## line 476 ##
0.0
sage: p.objective_coefficient(0, 2) ## line 478 ##
sage: p.objective_coefficient(0) ## line 479 ##
2
sage: sig_on_count() # check sig_on/off pairings (virtual doctest) ## line 481 ##
0
sage: from sage.numerical.backends.generic_backend import get_solver ## line 506 ##
sage: p = get_solver(solver="CVXPY") ## line 507 ##
sage: p.add_linear_constraints(5, 0, None) ## line 508 ##
sage: p.add_col(list(range(5)), list(range(5))) ## line 509 ##
sage: p.solve() ## line 513 ##
0
sage: p.objective_coefficient(0,1) ## line 515 ##
sage: p.solve() ## line 516 ##
```

But also some seemingly unrelated stuff:

```
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/doctest/test.py  # 2 doctests failed
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/interfaces/octave.py  # 1 doctest failed
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/numerical/backends/cvxpy_backend.pyx  # Killed due to segmentation fault
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/plot/plot3d/tachyon.py  # 2 doctests failed
sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/rings/integer.pyx  # 1 doctest failed
```

The GH CI also reports coverage-related issues, which I think will be fixed by rebasing the branch on `develop`

**Comment by Kerl13** [comment]
<div id="comment:37" align="right">comment:37</div>

I'm not familiar with the numerical backend logic yet, so it will take a bit more time to review.
Here are few first remarks on the new packages:

- cvxpy requires numpy to run, but also to build, shouldn't numpy be an order dependency as well?

- Same question for ecos and scs.

- The pyproject.toml file for cvxpy 1.2.1 specifies an upper bound for numpy that we don't comply with. I saw your comment in spkg-install.in about this. Are you sure it's okay to ignore it? What are the implication of setting `--no-build-isolation` there?

- Also, why adding this flag in osqp's `spkg-install.in` file as well? And ecos?

- qdldl requires pybind11 at build time, shouldn't it be an order dependency?

- cylp is released under EPL version 2 (not just EPL)
- More importantly, EPL (version 1 and 2) are incompatible with GPLv3 according to https://www.gnu.org/licenses/license-list.html#EPL2 . Cylp should be should explicitly double-licenced to be compatible with GPL.

**Comment by Kerl13** [comment]
<div id="comment:39" align="right">comment:39</div>

I'm not an expert of the numerical backend stuff, but here are a few remarks on the code.

Aside from that, I've been able to run some examples from https://doc.sagemath.org/html/en/thematic_tutorials/linear_programming.html without an issue. So I guess it looks good!

---

**file generic_backend.pyx, line 1650:**

```
from sage.numerical.backends.cvxpy_backend import CVXPYBackend
```
It is possible to avoid the `Unused entry 'CVXPYBackend'` cython warning using:

```
__import__("sage.numerical.backends.cvxpy_backend.CVXPYBackend")
```

**file generic_backend.pyx, line 1824:** the `partial` function is imported but unused

**In CVXPYBackend.__cinit__**: the arguments of the `__cinit__` method of CVXPYBackend are not documented. In particular, I have no idea what `cvxpy_solver_args` might be. Also I don't get why `base_ring` is an argument if the only legal value for it is `RDF`.

**file cvxpy_backend.pyx, line 114:**

```
import cvxpy as cp
cvxpy_solver = getattr(cp, cvxpy_solver)
```
cvxpy is already imported at the top of the module, you could replace these two lines by:

```
cvxpy_solver = getattr(cvxpy, cvxpy_solver)
```

**file cvxpy_backend.pyx, line 152, in `__copy__`:**
why `type(self)(base_ring=...)` and not simply `CVXPYBackend(base_ring=...)`.

**In cvxpy_backend.pyx**
I get the following cython warning:

```
[cython] FutureWarning: Cython directive 'language_level' not set, using 2 for now
(Py2). This will change in a later release!
```
Shouldn't we write `# cython: language_level = 3` at the top of every new cython file?

**file cvxpy_backend.pyx, line 188:**

```
- ``binary`` - ``True`` if the variable is binary (default: ``False``).

- ``continuous`` - ``True`` if the variable is binary (default: ``True``).

- ``integer`` - ``True`` if the variable is binary (default: ``False``).
```
This looks like the same docstring have been copy-pasted but not modified. This problem seems present in all the backends, I guess because it comes from the generic backend.

**CVXPYBackend.set_verbosity**: maybe the docstring for `set_verbosity` should state that this method is only here for compatibility and ignores its argument?

**CVXPYBackend.solve**: the `.solve` method in CVXPYBackend only has examples where solving fails, maybe it would be nice to have an example of normal usage of the function, where it succeeds?

**Comment by Kerl13** [comment]
Reviewer: **MartinPepin**

**Comment by mkoeppe** [comment]
<div id="comment:41" align="right">comment:41</div>

Replying to [Martin Pépin](#comment%3A37):
> I'm not familiar with the numerical backend logic yet, so it will take a bit more time to review.
> Here are few first remarks on the new packages:
> 
> - cvxpy requires numpy to run, but also to build, shouldn't numpy be an order dependency as well?

"order-only prerequisites" are strictly weaker than normal prerequisites, so one does not have to declare both. https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`7c05f49`](https://github.com/sagemath/sagetrac-mirror/commit/7c05f49ca0776b68edcad0b8277b4511b37d9eea)** to **[`0bb4309`](https://github.com/sagemath/sagetrac-mirror/commit/0bb4309013e34701fe773c9e44cdd2d5dd213a62)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:42"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/effa9dbeabf42cfd34757f05b8b2ce5176523258"><code>effa9db</code></a></td><td><code>Merge tag '9.7' into t/34251/add_optional_package_cvxpy</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0bb4309013e34701fe773c9e44cdd2d5dd213a62"><code>0bb4309</code></a></td><td><code>build/pkgs/cylp/SPKG.rst: Add more detailed license info</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:43" align="right">comment:43</div>

Replying to [Martin Pépin](#comment%3A37):
> - cylp is released under EPL version 2 (not just EPL)
> - More importantly, EPL (version 1 and 2) are incompatible with GPLv3 according to https://www.gnu.org/licenses/license-list.html#EPL2 . Cylp should be should explicitly double-licenced to be compatible with GPL.

Thanks, I've added something along these lines. We actually had a discussion on this problematic situation in #26511 not too long ago.

---
New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/effa9dbeabf42cfd34757f05b8b2ce5176523258"><code>effa9db</code></a></td><td><code>Merge tag '9.7' into t/34251/add_optional_package_cvxpy</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0bb4309013e34701fe773c9e44cdd2d5dd213a62"><code>0bb4309</code></a></td><td><code>build/pkgs/cylp/SPKG.rst: Add more detailed license info</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:44" align="right">comment:44</div>

Replying to [Martin Pépin](#comment%3A37):
> - The pyproject.toml file for cvxpy 1.2.1 specifies an upper bound for numpy that we don't comply with. I saw your comment in spkg-install.in about this. Are you sure it's okay to ignore it?

Note that the documentation https://github.com/cvxpy/cvxpy#installation does not mention such upper bounds.

In https://github.com/cvxpy/cvxpy/blob/master/pyproject.toml, they are following numpy's recommendations, which are (to my understanding) motivated by setting a standard for compatibility of wheels, by building platform wheels against the minimum supported numpy version for each Python version (and platform). This is irrelevant for the Sage distribution.

> What are the implication of setting `--no-build-isolation` there?

We sacrifice a small bit of build robustness that comes with the build isolation.
Using `--no-build-isolation` used to be the default for all packages in Sage until I changed it in #33789 (in the Sage 9.7 cycle).

**Comment by mkoeppe** [comment]
<div id="comment:45" align="right">comment:45</div>

Thanks for the careful reading of the docstrings! I'll make changes later today.

**Comment by mkoeppe** [comment]
Changed reviewer from **MartinPepin** to **Martin Pépin**

**Comment by mkoeppe** [comment]
<div id="comment:47" align="right">comment:47</div>

Replying to [Martin Pépin](#comment%3A39):
> **In cvxpy_backend.pyx**
> I get the following cython warning:
> 
> ```
> [cython] FutureWarning: Cython directive 'language_level' not set, using 2 for now
> (Py2). This will change in a later release!
> ```
> Shouldn't we write `# cython: language_level = 3` at the top of every new cython file?

Since #33544, we set this globally (in `src/sage_setup/cython_options.py`).

**Comment by mkoeppe** [comment]
<div id="comment:48" align="right">comment:48</div>

Replying to [Martin Pépin](#comment%3A39):
> **file cvxpy_backend.pyx, line 152, in `__copy__`:**
> why `type(self)(base_ring=...)` and not simply `CVXPYBackend(base_ring=...)`.

That's for the benefit of possible user-defined subclasses

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:49"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/944d112c2b036bccd7c878862a94d07288991721"><code>944d112</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Add doc for init args</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0bd13b165ad18a14b3761dc4a6404c88aee7b92e"><code>0bd13b1</code></a></td><td><code>src/sage/numerical/backends: Remove unnecessary imports, muffle unused variable warnings</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`0bb4309`](https://github.com/sagemath/sagetrac-mirror/commit/0bb4309013e34701fe773c9e44cdd2d5dd213a62)** to **[`0bd13b1`](https://github.com/sagemath/sagetrac-mirror/commit/0bd13b165ad18a14b3761dc4a6404c88aee7b92e)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`0bd13b1`](https://github.com/sagemath/sagetrac-mirror/commit/0bd13b165ad18a14b3761dc4a6404c88aee7b92e)** to **[`52d3f46`](https://github.com/sagemath/sagetrac-mirror/commit/52d3f46ce49607da6516f7f06e37ad0f6e558639)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:50"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/52d3f46ce49607da6516f7f06e37ad0f6e558639"><code>52d3f46</code></a></td><td><code>src/sage/numerical/backends: Fix copy-pasted typos in documentation (binary/continuous/integer)</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:51" align="right">comment:51</div>

Replying to [Martin Pépin](#comment%3A39):
> **In CVXPYBackend.__cinit__**: the arguments of the `__cinit__` method of CVXPYBackend are not documented. In particular, I have no idea what `cvxpy_solver_args` might be.

Thanks, I've added some documentation now.

> Also I don't get why `base_ring` is an argument if the only legal value for it is `RDF`.

Just an argument that all backend classes must support because `MixedIntegerLinearProgram` passes it.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`52d3f46`](https://github.com/sagemath/sagetrac-mirror/commit/52d3f46ce49607da6516f7f06e37ad0f6e558639)** to **[`b70559a`](https://github.com/sagemath/sagetrac-mirror/commit/b70559aa949ef0b7a2c6f28ffc71fd3f31fb1785)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:52"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b70559aa949ef0b7a2c6f28ffc71fd3f31fb1785"><code>b70559a</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Add to doc</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`b70559a`](https://github.com/sagemath/sagetrac-mirror/commit/b70559aa949ef0b7a2c6f28ffc71fd3f31fb1785)** to **[`c9f0b39`](https://github.com/sagemath/sagetrac-mirror/commit/c9f0b39aec965730bfae1fd4a0f1997423c7619f)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:53"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/63b84c31702f74d0263a5e3e0cec3add027f8421"><code>63b84c3</code></a></td><td><code>build/pkgs/scs/spkg-install.in: Use --no-build-isolation</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/c9f0b39aec965730bfae1fd4a0f1997423c7619f"><code>c9f0b39</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Update doctest output for bounds methods</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:54" align="right">comment:54</div>

Replying to [Martin Pépin](#comment%3A36):
> Some test do not pass, most notably: 
> 
> ```
> sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/numerical/backends/cvxpy_backend.pyx
>     Killed due to segmentation fault
> ```

OK, I can reproduce this now, using ubuntu-jammy

**Comment by Kerl13** [comment]
<div id="comment:55" align="right">comment:55</div>

Replying to [Matthias Köppe](#comment%3A41):
> "order-only prerequisites" are strictly weaker than normal prerequisites, so one does not have to declare both. https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

Oh I ignored it was actually a GNU make feature. Thanks for the pointer.

**Comment by Kerl13** [comment]
<div id="comment:56" align="right">comment:56</div>

Replying to [Matthias Köppe](#comment%3A43):
> We actually had a discussion on this problematic situation in #26511 not too long ago.

So the conclusion of the discussion is that it's okay to package EPL code as long as it's only provided as an alternative to another tool and thus is not used by default by sage?

**Comment by mkoeppe** [comment]
<div id="comment:57" align="right">comment:57</div>

Yes, Sage-the-distribution has the same role as a Linux distribution like Debian.

**Comment by Kerl13** [comment]
Attachment: **[cvxpy.log](https://github.com/sagemath/sage-prod/files/10659642/cvxpy.log)**

**Comment by Kerl13** [comment]
<div id="comment:58" align="right">comment:58</div>

After rebasing on develop, I now get the attached errors when running `sage -t --long --random-seed=262306895609334379547192037378099413502 src/sage/numerical/backends/cvxpy_backend.pyx`

Don't know why the segfault occurred before but the new errors look minor now

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`c9f0b39`](https://github.com/sagemath/sagetrac-mirror/commit/c9f0b39aec965730bfae1fd4a0f1997423c7619f)** to **[`539c970`](https://github.com/sagemath/sagetrac-mirror/commit/539c9705baef065d773f8a5349096679e6248e48)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:59"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/55fc4fbbcbaf437977a87b55829c406c7a6ffd02"><code>55fc4fb</code></a></td><td><code>build/pkgs/{cvxpy,scs,osqp_python,ecos_python,qdldl_python}: Add distros/conda.txt</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/18e7434763226568a693d0702924bd47ab7c8b60"><code>18e7434</code></a></td><td><code>build/pkgs/osqp_python/dependencies: Add cmake</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/11e4de8d045ee27cfdb8439fe84f5a5de67d1fd9"><code>11e4de8</code></a></td><td><code>build/pkgs/ecos_python/dependencies: Remove suitesparse</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/9a1eeec2c8cef1f1e7daf41e7b1f21aff92591e7"><code>9a1eeec</code></a></td><td><code>build/pkgs/cylp/SPKG.rst: Add more detailed license info</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b9c92aec7237099e0b31120cc06bb3335c4e523c"><code>b9c92ae</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Add doc for init args</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5b840bddd970e89247a6f17625cdd59c959a2c69"><code>5b840bd</code></a></td><td><code>src/sage/numerical/backends: Remove unnecessary imports, muffle unused variable warnings</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7132302e349532f02a7eb174114c3dcac5808250"><code>7132302</code></a></td><td><code>src/sage/numerical/backends: Fix copy-pasted typos in documentation (binary/continuous/integer)</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b5cbc5acda3505f8aa31b53052d6b21391a35498"><code>b5cbc5a</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Add to doc</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/83ce468ea3d852f4f3b1131c512d3380a82ab27b"><code>83ce468</code></a></td><td><code>build/pkgs/scs/spkg-install.in: Use --no-build-isolation</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/539c9705baef065d773f8a5349096679e6248e48"><code>539c970</code></a></td><td><code>src/sage/numerical/backends/cvxpy_backend.pyx: Update doctest output for bounds methods</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`539c970`](https://github.com/sagemath/sagetrac-mirror/commit/539c9705baef065d773f8a5349096679e6248e48)** to **[`7346fd1`](https://github.com/sagemath/sagetrac-mirror/commit/7346fd1e3c2f7ebf4c6b01090dd14bc87b1641c5)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:60"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/98c38c1743d3477628b76d27a15b64389e698dc8"><code>98c38c1</code></a></td><td><code>Merge tag '9.8.beta3' into t/34251/add_optional_package_cvxpy</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7211f4208965f7dd2c775be7fbf89e589f21fecd"><code>7211f42</code></a></td><td><code>build/pkgs/cvxpy: Update to 1.2.2</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7346fd1e3c2f7ebf4c6b01090dd14bc87b1641c5"><code>7346fd1</code></a></td><td><code>build/pkgs/scs: Update to 3.2.2</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:61" align="right">comment:61</div>

Tests at https://github.com/mkoeppe/sage/actions/runs/3476460100, https://github.com/mkoeppe/sage/actions/runs/3476460090

**Comment by sheerluck** [comment]
<div id="comment:63" align="right">comment:63</div>

we need this ticket for #31962

we need #31962 for #32053

**Comment by mkoeppe** [comment]
<div id="comment:64" align="right">comment:64</div>

We'll have to merge this with #21003, which touches the same files and will likely be in the next beta

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`7346fd1`](https://github.com/sagemath/sagetrac-mirror/commit/7346fd1e3c2f7ebf4c6b01090dd14bc87b1641c5)** to **[`666e973`](https://github.com/sagemath/sagetrac-mirror/commit/666e973a8de6cbf1cc4622f385320559efd4a960)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:65"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/666e973a8de6cbf1cc4622f385320559efd4a960"><code>666e973</code></a></td><td><code>Merge tag '9.8.beta7' into t/34251/add_optional_package_cvxpy</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`666e973`](https://github.com/sagemath/sagetrac-mirror/commit/666e973a8de6cbf1cc4622f385320559efd4a960)** to **[`7ff5766`](https://github.com/sagemath/sagetrac-mirror/commit/7ff5766056af2607d9f42fc8d04cafe215379ba5)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:66"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ba9fcbc817193edb7e443519ba8293111d3b0866"><code>ba9fcbc</code></a></td><td><code>build/pkgs/cvxpy: Update to 1.3.0</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/75155645dd1d1514bc57be9a7174923c9a1ec803"><code>7515564</code></a></td><td><code>build/pkgs/osqp_python: Update to 0.6.2.post8</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/91cbc216435076022488da4d20a26a9d160b56c2"><code>91cbc21</code></a></td><td><code>build/pkgs/ecos_python: Update to 2.0.12</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7ff5766056af2607d9f42fc8d04cafe215379ba5"><code>7ff5766</code></a></td><td><code>build/pkgs/qdldl_python: Update to 0.1.5.post3</code></td></tr></table>

**Comment by sheerluck** [comment]
<div id="comment:68" align="right">comment:68</div>

without this ticket I can do this

```
sage: from sage.combinat.designs.database import OA_10_469
sage: OA = OA_10_469()
sage: print(len(OA))
219961
```
this ticket just hangs (probably O(n<sup>2</sup>) while `p.set_objective(p.sum(b[fe] for fe in L))`, looks like every `b[fe]` creates new `cvxpy.Problem` with 1, 2, 3 ... 900 constraints)

**Comment by mkoeppe** [comment]
<div id="comment:69" align="right">comment:69</div>

Yes, we should definitely improve this

**Comment by dimpase** [comment]
<div id="comment:70" align="right">comment:70</div>

Replying to [@sheerluck](#comment%3A68):
> without this ticket I can do this
> 
> ```
> sage: from sage.combinat.designs.database import OA_10_469
> sage: OA = OA_10_469()
> sage: print(len(OA))
> 219961
> ```
> this ticket just hangs (probably O(n<sup>2</sup>) while `p.set_objective(p.sum(b[fe] for fe in L))`, looks like every `b[fe]` creates new `cvxpy.Problem` with 1, 2, 3 ... 900 constraints)

I don't see where in `OA_10_469()` any optimisation problem is set up. It seems to be a simple
combinatorial construction, no solvers involved. `OA` is just a list.

**Comment by sheerluck** [comment]
<div id="comment:71" align="right">comment:71</div>

Replying to [Dima Pasechnik](#comment%3A70):
> I don't see where in `OA_10_469()` any optimisation problem is set up.


```
database.py:1693 -> 
    matrix = _reorder_matrix(symmetric_design)

orthogonal_arrays_build_recursive.py:1189 ->
    matching = g.matching(algorithm="LP")

graph.py:4210 ->
    p = MixedIntegerLinearProgram(maximization=True, solver=solver)
```

**Comment by sheerluck** [comment]
and now I can link [a range of lines](https://github.com/sagemath/sage/blob/develop/src/sage/graphs/graph.py#L4205-L4215)

**Comment by mkoeppe** [comment]
Removed branch from issue description; replaced by PR #35120

**Comment by mkoeppe** [comment]
> this ticket just hangs (probably O(n2) while `p.set_objective(p.sum(b[fe] for fe in L))`, looks like every `b[fe]` creates new `cvxpy.Problem` with 1, 2, 3 ... 900 constraints)

I'd suggest that we improve performance in follow-up PRs
